bindings/nodejs/src/functions/i2c.cc

```diff
@@ -0,0 +1,529 @@
+        Nan::New((int)status)
+    };
+
+    callback->Call(2, arguments);
+    delete callback;
+    hijack_unref();
+}
+
+NAN_METHOD(bind_sol_i2c_write)
+{
+    VALIDATE_ARGUMENT_COUNT(info, 3);
+    VALIDATE_ARGUMENT_TYPE(info, 0, IsObject);
+    VALIDATE_ARGUMENT_TYPE(info, 1, IsObject);
+    VALIDATE_ARGUMENT_TYPE(info, 2, IsFunction);
+
+    uint8_t *outputBuffer = (uint8_t *) 0;
+    size_t count = 0;
+
+    Local<Object> jsI2c = Nan::To<Object>(info[0]).ToLocalChecked();
+    sol_i2c *i2c = (sol_i2c *)SolI2c::Resolve(jsI2c);
+
+    if (!node::Buffer::HasInstance(info[1])) {
+        Nan::ThrowTypeError("Argument 1 must be a node Buffer");
+        return;
+    }
+
+    count = node::Buffer::Length(info[1]);
+
+    outputBuffer = (uint8_t *) calloc(count, sizeof(uint8_t));
+    if (!outputBuffer)
+    {
+        Nan::ThrowError("Failed to allocate memory for output buffer");
+        return;
+    }
+
+    memcpy(outputBuffer, node::Buffer::Data(info[1]), count);
+
+    if (!hijack_ref()) {
+        free(outputBuffer);
+        return;
+    }
+
+    Nan::Callback *callback =
+            new Nan::Callback(Local<Function>::Cast(info[2]));
+    sol_i2c_pending *i2c_pending =
+            sol_i2c_write(i2c, outputBuffer, count, sol_i2c_write_cb, callback);
+
+    if (!i2c_pending) {
+        free(outputBuffer);
+        delete callback;
+        hijack_unref();
+        return;
+    }
+
+    info.GetReturnValue().Set(SolI2c::New(i2c_pending));
+}
+
+static void sol_i2c_write_reg_cb(void *cb_data, struct sol_i2c *i2c,
+                                 uint8_t reg, uint8_t *data, ssize_t status)
+{
+    Nan::HandleScope scope;
+    Nan::Callback *callback = (Nan::Callback *)cb_data;
+    Local<Value> buffer;
+
+    if (status >= 0) {
+        buffer = Nan::NewBuffer((char *)data, status).ToLocalChecked();
+    } else {
+        buffer = Nan::Null();
+        free(data);
+    }
+
+    Local<Value> arguments[3] = {
+        Nan::New(reg),
+        buffer,
+        Nan::New((int)status)
+    };
+
+    callback->Call(3, arguments);
+
+    delete callback;
+    hijack_unref();
+}
+
+NAN_METHOD(bind_sol_i2c_write_register)
+{
+    VALIDATE_ARGUMENT_COUNT(info, 4);
+    VALIDATE_ARGUMENT_TYPE(info, 0, IsObject);
+    VALIDATE_ARGUMENT_TYPE_OR_NULL(info, 1, IsUint32);
+    VALIDATE_ARGUMENT_TYPE_OR_NULL(info, 2, IsObject);
+    VALIDATE_ARGUMENT_TYPE(info, 3, IsFunction);
+
+    uint8_t *outputBuffer = (uint8_t *) 0;
+    size_t count;
+
+    Local<Object> jsI2c = Nan::To<Object>(info[0]).ToLocalChecked();
+    sol_i2c *i2c = (sol_i2c *)SolI2c::Resolve(jsI2c);
+
+    uint8_t reg = info[1]->Uint32Value();
+
+    if (!node::Buffer::HasInstance(info[2])) {
+        Nan::ThrowTypeError("Argument 2 must be a node Buffer");
+        return;
+    }
+
+    count = node::Buffer::Length(info[2]);
+
+    outputBuffer = (uint8_t *) calloc(count, sizeof(uint8_t));
+    if (!outputBuffer) {
+        Nan::ThrowError("Failed to allocate memory for output buffer");
```

Comment:
I've found curious/confusing that both read and write operations talk about 'output buffer'.